File: src/main/java/com/example/pi4micronaut/LabAutomation.java
package com.example.pi4micronaut;

import com.opensourcewithslu.inputdevices.*;
import com.opensourcewithslu.outputdevices.*;
import com.opensourcewithslu.utilities.MultipinConfiguration;


import com.pi4j.context.Context;

import com.pi4j.io.gpio.digital.DigitalInput;
import com.pi4j.io.i2c.I2CConfig;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import jakarta.inject.Named;

import java.lang.Thread;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@Controller("/LabAutomation")
public class LabAutomation {
    private final LCD1602Helper lcd;
    private final RGBLEDHelper rgb;
    private final PIRSensorHelper pirStart;
    private final PIRSensorHelper pirEnd;

    private Thread timer;
    private boolean inMaze;
    private int timeInMaze;
    private long startTime;
    private long endTime;

    private static final Logger log = LoggerFactory.getLogger(LabAutomation.class);

    public LabAutomation(
            @Named("lcd") I2CConfig lcd,
            @Named("rgb-led") MultipinConfiguration rgb,
            @Named("pir-sensor-1") DigitalInput pirStart,
            @Named("pir-sensor-2") DigitalInput pirEnd,
            Context pi4j)
    {
        this.lcd = new LCD1602Helper(lcd,pi4j);
        this.rgb = new RGBLEDHelper(rgb);
        this.pirStart = new PIRSensorHelper(pirStart);
        this.pirEnd = new PIRSensorHelper(pirEnd);

        this.inMaze = false;
        this.timeInMaze = 0;

        this.timer = new Thread(() -> {

            while ( true ) {

                try { Thread.sleep(50); }
                catch (InterruptedException e) { log.error("cant sleep :(", e); }

                if (!this.inMaze ) { continue; }

                this.lcd.writeTextAtLine("Current Time:", 1);
                this.lcd.writeTextAtLine(String.format("%d seconds", this.timeInMaze), 2);
                this.timeInMaze++;
                this.log.info("Mouse in Maze");

                try { Thread.sleep(950); }
                catch (InterruptedException e) { log.error("cant sleep :(", e); }
            }
        });
        this.timer.start();

    }

    /*
        For the Lab automation, we are going to assume the mouse maze.
        The idea of this maze is to figure out how much time the mouse takes to exit the maze.
        We are going to use 2 ultrasonic sensors at the entry and exit locations.
        - Start and end time need to be noted based on the ultrasonic sensor's input( should have a threshold of 4 CM).
        - Subtract end time with the start time to get the time taken by the mouse to exit.
        - Use LCD to display messages like Mouse entered, mouse left and total time
     */

    @Get("/enable")
    public void enable() {
        this.pirStart.addEventListener((event) -> {

            if ( this.pirStart.isMoving  && !this.inMaze ) {
                this.timeInMaze = 0;
                this.inMaze = true;
                this.rgb.setGreen(255);
                this.rgb.setRed(0);
                this.startTime = System.currentTimeMillis();
                this.lcd.clearDisplay();
            }

        });

        this.pirEnd.addEventListener((event) -> {

            if ( this.pirEnd.isMoving && this.inMaze ) {
                this.inMaze = false;
                this.rgb.setGreen(0);
                this.rgb.setRed(255);
                this.endTime = System.currentTimeMillis();
                long finishTime = ( this.endTime - this.startTime ) / 1000;
                this.lcd.writeTextAtLine("Final Time:", 1);
                this.lcd.writeTextAtLine(String.format("%d seconds", finishTime), 2);
            }
        });
    }
}

File: src/main/kotlin/com/labAutomation/LabAutomation.kt
package com.labAutomation

import com.opensourcewithslu.inputdevices.PIRSensorHelper
import com.opensourcewithslu.outputdevices.LCD1602Helper
import com.opensourcewithslu.outputdevices.RGBLEDHelper
import com.opensourcewithslu.utilities.MultipinConfiguration
import com.pi4j.context.Context
import com.pi4j.io.gpio.digital.DigitalInput
import com.pi4j.io.gpio.digital.DigitalStateChangeEvent
import com.pi4j.io.i2c.I2CConfig
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Get
import jakarta.inject.Named
import org.slf4j.Logger
import org.slf4j.LoggerFactory


@Controller("/LabAutomation")
class LabAutomation(
    @Named("lcd") lcd: I2CConfig?,
    @Named("rgb-led") rgb: MultipinConfiguration?,
    @Named("pir-sensor-1") pirStart: DigitalInput?,
    @Named("pir-sensor-2") pirEnd: DigitalInput?,
    pi4j: Context?
) {
    private val lcd = LCD1602Helper(lcd, pi4j)
    private val rgb = RGBLEDHelper(rgb)
    private val pirStart = PIRSensorHelper(pirStart)
    private val pirEnd = PIRSensorHelper(pirEnd)

    private val timer: Thread
    private var inMaze: Boolean
    private var startTime: Long = 0
    private var currentTime: Long = 0
    private var endTime: Long = 0

    // Lab Automation System detects when a mouse enters and leaves a maze and
    // displays the time it takes for the mouse to complete the maze in seconds.
    init {
        this.inMaze = false

        // Timer thread: counts and displays how long the mouse has been in the maze in seconds
        this.timer = Thread {
            while (true) {
                try {
                    Thread.sleep(100)
                } catch (e: InterruptedException) {
                    log.error("cant sleep :(", e)
                }

                // If the mouse is not in the maze, dont execute the rest of the while block
                if (!this.inMaze) {
                    continue
                }

                // Calculate how long the mouse has been in the maze and write to LCD display
                val newTime = System.currentTimeMillis()
                if (newTime - this.currentTime >= 1000) {
                    val displayTime = (newTime - this.startTime) / 1000
                    this.currentTime = newTime
                    this.lcd.writeTextAtLine("Current Time:", 1)
                    this.lcd.writeTextAtLine(String.format("%d seconds", displayTime), 2)
                }
            }
        }
        timer.start()
    }

    // Enables the Lab Automation system
    @Get("/enable")
    fun enable() {
        // Adds event listener for the motion sensor at the start of the maze

        pirStart.addEventListener { event: DigitalStateChangeEvent<*>? ->

            // Detects if mouse has entering the maze
            if (pirStart.isMoving && !this.inMaze) {
                this.inMaze = true
                rgb.setGreen(255)
                rgb.setRed(0)
                this.startTime = System.currentTimeMillis()
                this.currentTime = this.startTime
                lcd.clearDisplay()
            }
        }

        // Adds event listener for the motion sensor at the end of the maze
        pirEnd.addEventListener { event: DigitalStateChangeEvent<*>? ->

            // Detects if mouse has exited the maze
            if (pirEnd.isMoving && this.inMaze) {
                this.inMaze = false
                rgb.setGreen(0)
                rgb.setRed(255)
                this.endTime = System.currentTimeMillis()
                val finishTime =
                    (this.endTime - this.startTime) / 1000
                lcd.writeTextAtLine("Final Time:", 1)
                lcd.writeTextAtLine(String.format("%d seconds", finishTime), 2)
                log.info(String.format("Final Time: %d seconds", finishTime))
            }
        }
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(LabAutomation::class.java)
    }
}